File: api/services/user_records.py
import collections
from django.db.models import Count
from record.models import StatusType


class RecordFilter(object):
    def __init__(self, status_type, category_id, category_id_set):
        '''
        :param status_type: record.models.StatusType
        :param category_id: int?
        :param category_id_set: bool
        :return:
        '''
        assert status_type is None or isinstance(status_type, StatusType)
        assert category_id != 0
        assert category_id_set or category_id is None
        self.status_type = status_type
        self.status_type_set = status_type is not None
        self.category_id = category_id
        self.category_id_set = category_id_set

    def without_status_type(self):
        if self.status_type_set:
            return RecordFilter(
                status_type=None,
                category_id=self.category_id,
                category_id_set=self.category_id_set,
            )
        return self

    def without_category_id(self):
        if self.category_id_set:
            return RecordFilter(
                status_type=self.status_type,
                category_id=None,
                category_id_set=False,
            )
        return self


def filter_counts(counts, record_filter):
    if record_filter.status_type_set:
        counts = [row for row in counts if row['status_type'] == record_filter.status_type]
    if record_filter.category_id_set:
        counts = [row for row in counts if row['category_id'] == record_filter.category_id]
    return counts


def count(user, record_filter):
    counts = list(user.record_set.values('status_type', 'category_id').annotate(Count('id')))

    filtered_counts = filter_counts(counts, record_filter)

    count_by_status_type = collections.defaultdict(lambda: 0)
    count_by_status_type['_all'] = 0
    for row in filter_counts(counts, record_filter.without_status_type()):
        count_by_status_type['_all'] += row['id__count']
        count_by_status_type[StatusType(row['status_type']).name] += row['id__count']

    count_by_category_id = collections.defaultdict(lambda: 0)
    count_by_category_id['_all'] = 0
    for row in filter_counts(counts, record_filter.without_category_id()):
        count_by_category_id['_all'] += row['id__count']
        count_by_category_id[row['category_id'] or '0'] += row['id__count']

    return {
        'total': sum(row['id__count'] for row in counts),
        'filtered': sum(row['id__count'] for row in filtered_counts),
        'by_status_type': count_by_status_type,
        'by_category_id': count_by_category_id,
    }

File: backend-ts/src2/services/userRecords.ts
import { Record } from "src/entities/record.entity";
import { StatusType } from "src/entities/status_type";
import { User } from "src/entities/user.entity";
import { db } from "src2/database";

export type CountByCriteria = {[key: string]: number} & {_all: number}

export type RecordsCount = {
  total: number;
  filtered: number;
  by_status_type: CountByCriteria;
  by_category_id: CountByCriteria;
}

type CountGroupRow = {
  count: number;
  status_type: StatusType;
  category_id: number | null;
}

export async function countRecordsForFilter(
  user: User,
  {
    statusType,
    categoryId,
  }: {
    statusType: StatusType | null,
    categoryId: number | null,
  }
): Promise<RecordsCount> {
  const counts: CountGroupRow[] = (await db.createQueryBuilder(Record, 'r')
    .select('COUNT(*)', 'count')
    .addSelect('status_type')
    .addSelect('category_id')
    .where({ user })
    .groupBy('status_type')
    .addGroupBy('category_id')
    .getRawMany())
    .map(it => ({
      count: Number(it.count), // count(*) returned as string when getRawMany
      status_type: it.status_type,
      category_id: it.category_id,
    }))
  const filtered = filterCounts(counts, statusType, categoryId)
  const countByStatusType: CountByCriteria = { _all: 0 }
  for (const row of filterCounts(counts, null, categoryId)) {
    countByStatusType._all += row.count
    const key = StatusType[row.status_type].toLowerCase()
    if (!countByStatusType[key]) countByStatusType[key] = 0
    countByStatusType[key] += row.count
  }
  const countByCategory: CountByCriteria = { _all: 0 }
  for (const row of filterCounts(counts, statusType, null)) {
    countByCategory._all += row.count
    const key = row.category_id?.toString() ?? '0'
    if (!countByCategory[key]) countByCategory[key] = 0
    countByCategory[key] += row.count
  }
  return {
    total: counts.reduce((acc, it) => acc + it.count, 0),
    filtered: filtered.reduce((acc, it) => acc + it.count, 0),
    by_status_type: countByStatusType,
    by_category_id: countByCategory,
  }
}

function filterCounts(
  counts: CountGroupRow[],
  statusType: StatusType | null,
  categoryId: number | null
): CountGroupRow[] {
  let result = counts
  if (statusType != null) {
    result = result.filter(it => it.status_type === statusType)
  }
  if (categoryId != null) {
    result = result.filter(it => (it.category_id ?? 0) === categoryId)
  }
  return result
}
